rfd/0226-self-healing.md

```diff
@@ -0,0 +1,286 @@
+not automatically reconnect in this scenario. Within gRPC client connections are
+maintained by load balancing policies[^2].
+
+Evaluating the default `pick_first` policy. It will only create a new connection
+when the existing TCP connection is disrupted. When health checking is enabled
+the client begins failing requests immediately rather than sending them to the
+server but never attempts to reconnect.
+
+Evaluating the `round_robin` policy, a connection is created to each address that
+was returned by the resolver. This policy is able to send requests to healthy
+connections based on health checks. However this assumes that each address represents
+a unique server instance. This is not usually the case. Teleport is typically fronted by one or more layers of load balancers. This means there could be
+a single address or connecting to multiple addresses could result in connections
+to the same server instance.
+
+In summary, neither the `pick_first` or the `round_robin` load balancing policies
+will work for our use case. Other built-in policies are available but they also
+do not meet our use case[^3].
+
+The solution here is to implement our own load balancing policy. The go-grpc library
+conveniently allows you to register custom policies[^2] by implementing the
+balancer interface[^4]. This policy will enable gRPC health checking by default
+and use the streaming API to ensure health checks can scale to a large number of
+clients.
+
+To avoid introducing too much new behavior we will base our implementation off of
+the existing `pick_first` policy[^5]. The primary difference will be when the existing
+connection enters the `TRANSIENT_FAILURE` state[^6] new connection attempts will be
+made until either the existing connection returns to `READY` or the new connection
+reaches the `READY` state.
+
+All connection attempts should use a backoff policy with jitter to avoid a thundering
+heard problem.
+
+If the new connection reaches the `READY` state it will replace the old connection
+and the old connection will be shutdown.
+
+We will differentiate between a connection failing due to the service being unhealthy
+and a connection failing due to a network issue.
+
+A connection failing due to a network issue can simply create a new conneciton.
+
+A connection failing due to an unhealthy service should continue to send RPCs to the
+unhealthy service until a new connection to a healthy service can be established.
+
+This allows for a partially degraded auth service to attempt to service RPCs. This
+preservers todays behavior where all requests get sent to the auth regardless of its
+health.
+
+Once a new connection is established the old connection will be drained gracefully
+and new RPCs will be sent over the new connection.
+
+Any open RPCs on the old connection will be allowed to run to completetion.
+
+Clients will discover the new load balancing policy via the `/webapi/ping` endpoint.
```

Comment:
I was talking with @sclevine and he suggested using `/webapi/find` over `/webapi/ping` since find is a cached result, while ping hits auth each time.